api/src/Security/Voter/CampRoleVoter.php

```diff
@@ -48,12 +51,20 @@ protected function voteOnAttribute(string $attribute, mixed $subject, TokenInter
             return true;
         }
 
-        return $camp->collaborations
+        $campCollaboration = $camp->collaborations
             ->filter(self::withStatus(CampCollaboration::STATUS_ESTABLISHED))
             ->filter(self::ofUser($user))
             ->filter(self::withRole($attribute))
-            ->exists(fn () => true)
+            ->first()
         ;
+
+        if ($campCollaboration) {
```

Comment:
Could we extract this tagging somehow into a composition pattern, instead of spilling caching logic into our voter classes directly? I think once we start implementing live updates, we will need a similar tagging mechanism, and I'd prefer to keep this stuff separate and pluggable if possible.

Reply:
Not sure I understand the idea. Can you elaborate a bit more on the proposal?

Regarding live updates: Yes, I can hope we can reuse a lot of the concept. Because invalidating a cache entry is really just a special form of a live update (in this case just only to the reverse proxy instead of directly to the client).

Reply:
Well, Symfony loves to work with composition over inheritance. So we could create a new class `CachingCampRoleVoter` which takes as a constructor argument a "normal" `CampRoleVoter $inner`.
In the `voteOnAttribute` method of the CachingCampRoleVoter, it would mostly call `$inner->voteOnAttribute`, and additionally add tags to the responseTagger as needed. So the tag detection and adding logic would be encapsulated in the `CachingCampRoleVoter`, while the vanilla voter logic would remain more or less untouched in the `CampRoleVoter`. If we come around to adding live updates, that might become another layer wrapping around the voter and delegating to the inner implementations, etc.

Reply:
I think that would lead to quite some code duplication, no?

For the `CachingCampRoleVoter` to work, it's not sufficient to know, whether `CampRoleVoter` voted true/false. It also needs to now the `$campCollaboration` which was responsible for a positive vote.

I see this delegation pattern works nicely with the Serializer, where each layer is independent of each other and still works if one layer is enabled/disabled. But here in this case, the functionality of adding tags is tightly coupled with the voting functionality itself.

Let's discuss in the upcoming core meeting. Might be easier to discuss verbally.